Mio build broken on Haiku, no kpoll/kqueue

```
   Compiling parking_lot v0.9.0
   Compiling hyper v0.12.35
   Compiling aho-corasick v0.7.15
   Compiling addr2line v0.14.0
   Compiling idna v0.1.5
   Compiling idna v0.2.0
   Compiling mime_guess v2.0.3
   Compiling futures-cpupool v0.1.8
   Compiling bytes v0.4.12
   Compiling mio v0.6.23
error[E0432]: unresolved imports `self::unix::Events`, `self::unix::Selector`
 --> /boot/home/.cargo/registry/src/github.com-1ecc6299db9ec823/mio-0.6.23/src/sys/mod.rs:5:5
  |
5 |     Events,
  |     ^^^^^^ no `Events` in `sys::unix`
6 |     Io,
7 |     Selector,
  |     ^^^^^^^^ no `Selector` in `sys::unix`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0432`.
error: could not compile `mio`
```

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -162,6 +162,7 @@ jobs:
           - x86_64-pc-windows-msvc
           - x86_64-unknown-dragonfly
           - x86_64-unknown-freebsd
+          - x86_64-unknown-haiku
           - x86_64-unknown-hermit
           - x86_64-unknown-illumos
           - x86_64-unknown-linux-gnu

--- a/README.md
+++ b/README.md
@@ -152,11 +152,9 @@ This uses the Windows AFD system to access socket readiness events.
 
 ### Unsupported
 
-* Haiku, see [issue #1472]
 * Solaris, see [issue #1152]
 * Wine, see [issue #1444]
 
-[issue #1472]: https://github.com/tokio-rs/mio/issues/1472
 [issue #1152]: https://github.com/tokio-rs/mio/issues/1152
 [issue #1444]: https://github.com/tokio-rs/mio/issues/1444
 

--- a/src/poll.rs
+++ b/src/poll.rs
@@ -3,6 +3,7 @@
     not(mio_unsupported_force_poll_poll),
     not(any(
         target_os = "espidf",
+        target_os = "haiku",
         target_os = "hermit",
         target_os = "nto",
         target_os = "solaris",
@@ -438,6 +439,7 @@ impl Poll {
     not(mio_unsupported_force_poll_poll),
     not(any(
         target_os = "espidf",
+        target_os = "haiku",
         target_os = "hermit",
         target_os = "nto",
         target_os = "solaris",
@@ -735,6 +737,7 @@ impl fmt::Debug for Registry {
     not(mio_unsupported_force_poll_poll),
     not(any(
         target_os = "espidf",
+        target_os = "haiku",
         target_os = "hermit",
         target_os = "nto",
         target_os = "solaris",

--- a/src/sys/unix/mod.rs
+++ b/src/sys/unix/mod.rs
@@ -36,7 +36,7 @@ cfg_os_poll! {
 
     cfg_io_source! {
         // Both `kqueue` and `epoll` don't need to hold any user space state.
-        #[cfg(not(any(mio_unsupported_force_poll_poll, target_os = "espidf", target_os = "hermit", target_os = "nto", target_os = "solaris", target_os = "vita")))]
+        #[cfg(not(any(mio_unsupported_force_poll_poll, target_os = "espidf", target_os = "haiku", target_os = "hermit", target_os = "nto", target_os = "solaris", target_os = "vita")))]
         mod stateless_io_source {
             use std::io;
             use std::os::fd::RawFd;
@@ -88,10 +88,10 @@ cfg_os_poll! {
             }
         }
 
-        #[cfg(not(any(mio_unsupported_force_poll_poll, target_os = "espidf", target_os = "hermit", target_os = "nto", target_os = "solaris", target_os = "vita")))]
+        #[cfg(not(any(mio_unsupported_force_poll_poll, target_os = "espidf", target_os = "haiku", target_os = "hermit", target_os = "nto", target_os = "solaris", target_os = "vita")))]
         pub(crate) use self::stateless_io_source::IoSourceState;
 
-        #[cfg(any(mio_unsupported_force_poll_poll, target_os = "espidf", target_os = "hermit", target_os = "nto", target_os = "solaris", target_os = "vita"))]
+        #[cfg(any(mio_unsupported_force_poll_poll, target_os = "espidf", target_os = "haiku", target_os = "hermit", target_os = "nto", target_os = "solaris", target_os = "vita"))]
         pub(crate) use self::selector::IoSourceState;
     }
 
@@ -102,6 +102,7 @@ cfg_os_poll! {
         mio_unsupported_force_waker_pipe,
         target_os = "aix",
         target_os = "dragonfly",
+        target_os = "haiku",
         target_os = "illumos",
         target_os = "netbsd",
         target_os = "nto",

--- a/src/sys/unix/net.rs
+++ b/src/sys/unix/net.rs
@@ -105,14 +105,17 @@ pub(crate) fn socket_addr(addr: &SocketAddr) -> (SocketAddrCRepr, libc::socklen_
                 sin_family: libc::AF_INET as libc::sa_family_t,
                 sin_port: addr.port().to_be(),
                 sin_addr,
-                #[cfg(not(target_os = "vita"))]
+                #[cfg(not(any(target_os = "haiku", target_os = "vita")))]
                 sin_zero: [0; 8],
+                #[cfg(target_os = "haiku")]
+                sin_zero: [0; 24],
                 #[cfg(target_os = "vita")]
                 sin_zero: [0; 6],
                 #[cfg(any(
                     target_os = "aix",
                     target_os = "dragonfly",
                     target_os = "freebsd",
+                    target_os = "haiku",
                     target_os = "ios",
                     target_os = "macos",
                     target_os = "netbsd",
@@ -147,6 +150,7 @@ pub(crate) fn socket_addr(addr: &SocketAddr) -> (SocketAddrCRepr, libc::socklen_
                     target_os = "aix",
                     target_os = "dragonfly",
                     target_os = "freebsd",
+                    target_os = "haiku",
                     target_os = "ios",
                     target_os = "macos",
                     target_os = "netbsd",

--- a/src/sys/unix/pipe.rs
+++ b/src/sys/unix/pipe.rs
@@ -28,6 +28,7 @@ pub(crate) fn new_raw() -> io::Result<[RawFd; 2]> {
 
     #[cfg(any(
         target_os = "aix",
+        target_os = "haiku",
         target_os = "ios",
         target_os = "macos",
         target_os = "tvos",
@@ -61,6 +62,7 @@ pub(crate) fn new_raw() -> io::Result<[RawFd; 2]> {
         target_os = "android",
         target_os = "dragonfly",
         target_os = "freebsd",
+        target_os = "haiku",
         target_os = "illumos",
         target_os = "ios",
         target_os = "linux",

--- a/src/sys/unix/selector/mod.rs
+++ b/src/sys/unix/selector/mod.rs
@@ -23,6 +23,7 @@ pub(crate) use self::epoll::{event, Event, Events, Selector};
 #[cfg(any(
     mio_unsupported_force_poll_poll,
     target_os = "espidf",
+    target_os = "haiku",
     target_os = "hermit",
     target_os = "nto",
     target_os = "solaris",
@@ -33,6 +34,7 @@ mod poll;
 #[cfg(any(
     mio_unsupported_force_poll_poll,
     target_os = "espidf",
+    target_os = "haiku",
     target_os = "hermit",
     target_os = "nto",
     target_os = "solaris",
@@ -44,6 +46,7 @@ cfg_io_source! {
     #[cfg(any(
         mio_unsupported_force_poll_poll,
         target_os = "espidf",
+        target_os = "haiku",
         target_os = "hermit",
         target_os = "nto",
         target_os = "solaris",

--- a/src/sys/unix/tcp.rs
+++ b/src/sys/unix/tcp.rs
@@ -89,6 +89,7 @@ pub(crate) fn accept(listener: &net::TcpListener) -> io::Result<(net::TcpStream,
     // set `CLOEXEC`.
     #[cfg(any(
         target_os = "aix",
+        target_os = "haiku",
         target_os = "ios",
         target_os = "macos",
         target_os = "redox",

--- a/src/sys/unix/uds/listener.rs
+++ b/src/sys/unix/uds/listener.rs
@@ -35,6 +35,7 @@ pub(crate) fn accept(listener: &net::UnixListener) -> io::Result<(UnixStream, So
 
     #[cfg(not(any(
         target_os = "aix",
+        target_os = "haiku",
         target_os = "ios",
         target_os = "macos",
         target_os = "netbsd",
@@ -62,6 +63,7 @@ pub(crate) fn accept(listener: &net::UnixListener) -> io::Result<(UnixStream, So
 
     #[cfg(any(
         target_os = "aix",
+        target_os = "haiku",
         target_os = "ios",
         target_os = "macos",
         target_os = "netbsd",

--- a/src/sys/unix/uds/mod.rs
+++ b/src/sys/unix/uds/mod.rs
@@ -87,6 +87,7 @@ where
 {
     #[cfg(not(any(
         target_os = "aix",
+        target_os = "haiku",
         target_os = "ios",
         target_os = "macos",
         target_os = "nto",
@@ -110,6 +111,7 @@ where
     // there is an error, the file descriptors are closed.
     #[cfg(any(
         target_os = "aix",
+        target_os = "haiku",
         target_os = "ios",
         target_os = "macos",
         target_os = "nto",

--- a/src/sys/unix/waker.rs
+++ b/src/sys/unix/waker.rs
@@ -13,6 +13,7 @@
     )),
     not(any(
         target_os = "espidf",
+        target_os = "haiku",
         target_os = "hermit",
         target_os = "nto",
         target_os = "solaris",
@@ -74,6 +75,7 @@ mod fdbased {
     )),
     not(any(
         target_os = "espidf",
+        target_os = "haiku",
         target_os = "hermit",
         target_os = "nto",
         target_os = "solaris",
@@ -239,6 +241,7 @@ pub use self::kqueue::Waker;
     mio_unsupported_force_waker_pipe,
     target_os = "aix",
     target_os = "dragonfly",
+    target_os = "haiku",
     target_os = "illumos",
     target_os = "netbsd",
     target_os = "nto",
@@ -294,6 +297,7 @@ mod pipe {
         #[cfg(any(
             mio_unsupported_force_poll_poll,
             target_os = "espidf",
+            target_os = "haiku",
             target_os = "nto",
             target_os = "solaris",
             target_os = "vita",
@@ -335,6 +339,7 @@ mod pipe {
             target_os = "redox",
         )
     ),
+    target_os = "haiku",
     target_os = "nto",
     target_os = "solaris",
     target_os = "vita",
@@ -344,6 +349,7 @@ pub(crate) use self::pipe::WakerInternal;
 #[cfg(any(
     mio_unsupported_force_poll_poll,
     target_os = "espidf",
+    target_os = "haiku",
     target_os = "hermit",
     target_os = "nto",
     target_os = "solaris",
@@ -377,6 +383,7 @@ mod poll {
 #[cfg(any(
     mio_unsupported_force_poll_poll,
     target_os = "espidf",
+    target_os = "haiku",
     target_os = "hermit",
     target_os = "nto",
     target_os = "solaris",

